WakaTime/WakaTime.swift

```diff
@@ -66,6 +68,86 @@ struct WakaTime: App {
+        }
+        URLSession.shared.dataTask(with: request) { data, response, error in
+            guard let data, let httpResponse = response as? HTTPURLResponse else {
+                completion(nil, error)
+                return
+            }
+            if httpResponse.statusCode == 304 {
+                DispatchQueue.main.async {
+                    // Current version is still the latest version available
+                    completion(currentVersion, nil)
+                }
+            } else if let lastModified = httpResponse.value(forHTTPHeaderField: "Last-Modified"),
+                      let release = try? JSONDecoder().decode(Release.self, from: data) {
+                // Remote version successfully decoded
+                ConfigFile.setSetting(section: "internal", key: "cli_version_last_modified", val: lastModified, internalConfig: true)
+                ConfigFile.setSetting(section: "internal", key: "cli_version", val: release.tagName, internalConfig: true)
+                DispatchQueue.main.async {
+                    completion(release.tagName, nil)
+                }
+            } else {
+                DispatchQueue.main.async {
+                    // Unexpected response
+                    completion(nil, nil)
+                }
+            }
+        }.resume()
+    }
+    
+    private func isCLILatest(completion: @escaping (Bool) -> Void) {
+        let cli = NSString.path(withComponents: FileManager.default.homeDirectoryForCurrentUser.pathComponents + [".wakatime", "wakatime-cli"])
+        guard FileManager.default.fileExists(atPath: cli) else {
+            completion(false)
+            return
+        }
+        let outputPipe = Pipe()
+        let process = Process()
+        process.launchPath = cli
+        process.arguments = ["--version"]
+        process.standardOutput = outputPipe
+        process.standardError = FileHandle.nullDevice
+        do {
+            try process.run()
+        } catch {
+            // Error running CLI process
+            completion(false)
+            return
+        }
+        let outputData = outputPipe.fileHandleForReading.readDataToEndOfFile()
+        let output = String(decoding: outputData, as: UTF8.self)
+        let version = output.firstMatch(of: /([0-9]+\.[0-9]+\.[0-9]+)/)
+        getLatestVersion { remoteVersion, error in
+            guard let remoteVersion, error == nil else {
+                // Could not retrieve remote version
+                completion(true)
+                return
+            }
+            if let version, "v" + version.0 == remoteVersion {
+                // Local version up to date
+                completion(true)
+            } else {
+                // Newer version available
+                completion(false)
+            }
+        }
+    }
+    
     private func downloadCLI() {
         let dir = NSString.path(withComponents: FileManager.default.homeDirectoryForCurrentUser.pathComponents + [".wakatime"])
```

Comment:
One thing that I generally wasn't sure about is the async handling of `downloadCLI` or that lack thereof. Shouldn't we add some completion handler to make sure other code isn't relying on the CLI being installed while it is still downloading? I thought that might be beyond the scope of this issue and left it out for now.

Reply:
Also if `wakatime-cli` is executed but fails we should gracefully handle that, so it doesn't matter if we're not ready yet for now.